### PR TITLE
stylo: Bug 1392161 - Introduce CSSPixelLength as computed::Length

### PR DESCRIPTION
**Pull request by BorisChiou**
These are the inter-dependent patches of bug 1392161. We want to handle
extreme small lengths carefully for some properties, such as transform, so we
shouldn't use |Au| as the computed value of specified::Length. Now, we introduce
a new type, CSSPixelLength, which is a wrapper of CSSFloat, and it is the
computed value of specified::Length, so we can keep the fractional part
of computed::Length.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix [Bug 1392161](https://bugzilla.mozilla.org/show_bug.cgi?id=1392161).
- [X] These changes do not require tests because there is a wpt test for this, and I also add some others in Gecko.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18381)

### CODE REVIEW

**Comment by BorisChiou** [comment]
r?@emilio

Assume handling extremely small lengths for transform is necessary, so we should avoid using ```Au``` as the computed lengths of transform. The current implementation has many duplicated codes, but I don't have a good idea to reduce them. @nox has a good idea?

also cc @upsuper

**Comment by nox** [comment]
As long as it works it's fine I guess. I don't currently have the mental bandwidth to think about how to derive more of this.

File: components/style/values/specified/length.rs [hunk 1]
```diff
@@ -116,22 +149,31 @@ impl FontRelativeLength {
...
                     // https://drafts.csswg.org/css-values/#ex
                     //
                     //     In the cases where it is impossible or impractical to
                     //     determine the x-height, a value of 0.5em must be
                     //     assumed.
                     //
-                    FontMetricsQueryResult::NotAvailable => reference_font_size.scale_by(0.5 * length),
-                }
+                    FontMetricsQueryResult::NotAvailable => {
+                        FontReferenceSize::Half(reference_font_size)
```
Comment (emilio) [inline components/style/values/specified/length.rs]: Why is this needed? Can't we return just `reference_font_size * 0.5` here?

File: components/style/values/specified/length.rs [hunk 2]
```diff
@@ -144,12 +186,13 @@ impl FontRelativeLength {
                     //
                     FontMetricsQueryResult::NotAvailable => {
                         if context.style().writing_mode.is_vertical() {
-                            reference_font_size.scale_by(length)
+                            FontReferenceSize::Full(reference_font_size)
                         } else {
-                            reference_font_size.scale_by(0.5 * length)
+                            FontReferenceSize::Half(reference_font_size)
```
Comment (emilio) [inline components/style/values/specified/length.rs]: Ditto, I don't get the point of `FontReferenceSize`, care to explain?

File: components/style/values/specified/length.rs [hunk 3]
```diff
@@ -282,6 +331,31 @@ impl AbsoluteLength {
...
+    #[inline]
+    pub fn to_px(&self) -> CSSFloat {
+        use std::f32;
+
+        const PX_PER_IN: CSSFloat = 96.;
+        const PX_PER_CM: CSSFloat = PX_PER_IN / 2.54;
+        const PX_PER_MM: CSSFloat = PX_PER_IN / 25.4;
+        const PX_PER_Q: CSSFloat = PX_PER_MM / 4.;
+        const PX_PER_PT: CSSFloat = PX_PER_IN / 72.;
+        const PX_PER_PC: CSSFloat = PX_PER_PT * 12.;
```
Comment (emilio) [inline components/style/values/specified/length.rs]: Can we remove the `AU_PER_FOO` constants and at least reuse this? If not, why not?

File: components/style/values/specified/transform.rs [hunk 4]
```diff
@@ -8,6 +8,7 @@ use cssparser::Parser;
 use parser::{Parse, ParserContext};
 use selectors::parser::SelectorParseError;
 use style_traits::{ParseError, StyleParseError};
+use super::Either;
```
Comment (emilio) [inline components/style/values/specified/transform.rs]: Let's not use `super::Either`, just `values::Either`.

File: components/style/values/specified/transform.rs [hunk 5]
```diff
@@ -254,3 +255,39 @@ impl ToComputedValue for TimingFunction {
...
+/// lengths converted into absolute lengths. Therefore, we need to define a different
+/// LengthOrPercentage type for transform.
+// https://drafts.csswg.org/css-transforms/#transform-property
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Debug, PartialEq, ToCss)]
+pub struct TransformLengthOrPercentage(pub LengthOrPercentage);
+
+impl Parse for TransformLengthOrPercentage {
+    #[inline]
+    fn parse<'i, 't>(context: &ParserContext,
```
Comment (emilio) [inline components/style/values/specified/transform.rs]: Let's indent this properly.

File: components/style/values/specified/transform.rs [hunk 6]
```diff
@@ -254,3 +255,39 @@ impl ToComputedValue for TimingFunction {
...
+
+/// The Length type for transform function.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[derive(Clone, Debug, PartialEq, ToCss)]
+pub struct TransformLength(pub Length);
+
+impl Parse for TransformLength {
+    #[inline]
+    fn parse<'i, 't>(context: &ParserContext,
+                     input: &mut Parser<'i, 't>) -> Result<Self, ParseError<'i>> {
```
Comment (emilio) [inline components/style/values/specified/transform.rs]: ditto, let's use block indentation here.

File: components/style/values/computed/transform.rs [hunk 7]
```diff
@@ -170,3 +164,171 @@ impl TransformList {
         }
     }
 }
+
+/// The computed value of TransformLengthOrPercentage. We treat computed values of transform
+/// as the specified values, but need to convert the relative lengths into absolute lengths.
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+#[animate(fallback = "Self::animate_fallback")]
+#[derive(Animate, Clone, Copy, Debug, PartialEq, ToAnimatedZero, ToCss)]
+pub enum TransformLengthOrPercentage {
```
Comment (emilio) [inline components/style/values/computed/transform.rs]: This is still a lot of duplicated code :(

**Review by emilio — COMMENTED** [review]
After seeing the other changes in https://bugzilla.mozilla.org/show_bug.cgi?id=1396692, I think the right way to do this is just using `CSSFloat` as the computed value everywhere (maybe with a new type, `CSSPixel`?), instead of duplicating all this code.

That'd solve the precision problem everywhere, is there any reason that can't be done?

**Comment by nox** [comment]
I agree with @emilio. Maybe we don't need to fix the precision problem everywhere indeed, but if fixing it everywhere means we have less code to maintain, I'm all for it.

**Comment by BorisChiou** [comment]
Thanks emilio. Thanks. it makes sense, and that's the idea way to do this. At first, I'm not sure the main purpose of using ```Au```. Thanks for clarification. I will land https://bugzilla.mozilla.org/show_bug.cgi?id=1396692 first (because it is much simpler), and then do replace Length with CSSPixelLength. :)

File: components/style/values/specified/length.rs (hunk 1 above)
Comment (BorisChiou) [reply components/style/values/specified/length.rs]: Yes. Thanks. Actually we don't really need it. I just want to use a type to represent the reference size. I will drop it.

**Comment by BorisChiou** [comment]
gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=57ce027c52b4

File: components/style/values/specified/length.rs [hunk 8]
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
                 (length, cmp::max(viewport_size.width, viewport_size.height)),
         };
 
-        // See bug 989802. We truncate so that adding multiple viewport units
-        // that add up to 100 does not overflow due to rounding differences
-        let trunc_scaled = ((length.0 as f64) * factor as f64 / 100.).trunc();
-        Au::from_f64_au(trunc_scaled)
+        // FIXME: Find a better way to drop the if-condition.
```
Comment (BorisChiou) [inline components/style/values/specified/length.rs]: I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

File: components/style/values/specified/length.rs [hunk 9]
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
...
+        let au = length.0 as f64 * factor as f64 / 100.;
+        if au > 1. {
+            // See bug 989802. We truncate so that adding multiple viewport units
+            // that add up to 100 does not overflow due to rounding differences
+            CSSPixelLength::from(Au::from_f64_au(au.trunc()))
+        } else {
+            // For the app unit value smaller then 1.0, we should keep its fraction part,
+            // So we don't truncate it.
+            CSSPixelLength::new((au / AU_PER_PX as f64) as f32)
```
Comment (BorisChiou) [inline components/style/values/specified/length.rs]: I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

File: components/style/values/specified/length.rs [hunk 10]
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
...
-        // that add up to 100 does not overflow due to rounding differences
-        let trunc_scaled = ((length.0 as f64) * factor as f64 / 100.).trunc();
-        Au::from_f64_au(trunc_scaled)
+        // FIXME: Find a better way to drop the if-condition.
+        let au = length.0 as f64 * factor as f64 / 100.;
+        if au > 1. {
+            // See bug 989802. We truncate so that adding multiple viewport units
+            // that add up to 100 does not overflow due to rounding differences
+            CSSPixelLength::from(Au::from_f64_au(au.trunc()))
+        } else {
```
Comment (BorisChiou) [inline components/style/values/specified/length.rs]: I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

File: components/style/values/specified/length.rs [hunk 11]
```diff
@@ -209,10 +234,17 @@ impl ViewportPercentageLength {
...
-        // See bug 989802. We truncate so that adding multiple viewport units
-        // that add up to 100 does not overflow due to rounding differences
-        let trunc_scaled = ((length.0 as f64) * factor as f64 / 100.).trunc();
-        Au::from_f64_au(trunc_scaled)
+        // FIXME: Find a better way to drop the if-condition.
+        let au = length.0 as f64 * factor as f64 / 100.;
+        if au > 1. {
+            // See bug 989802. We truncate so that adding multiple viewport units
+            // that add up to 100 does not overflow due to rounding differences
+            CSSPixelLength::from(Au::from_f64_au(au.trunc()))
```
Comment (BorisChiou) [inline components/style/values/specified/length.rs]: I added this during rebasing my patch. Manish fixed this for some rounding issues of viewport, but I need more time to check how to combine his patch with mine, so use this if-condition as a work-around for now.

**Comment by BorisChiou** [comment]
If gecko try looks good, I will land this ASAP.

File: components/style/values/specified/length.rs (hunk 11 above)
Comment (BorisChiou) [reply components/style/values/specified/length.rs]: OK, it seems this break other test, so I should update this now. :(

**Comment by BorisChiou** [comment]
Gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=e930cba9ad80
